File: sorting_algo.py
def merge_sort(array: list[int]) -> list[int]:
    """Sort a given list of int using merge sort algorithm
    Source: https://www.programiz.com/dsa/merge-sort

    Args:
        array (int): An array that want to be sorted
    """
    if len(array) > 1:

        #  mid is the point where the array is divided into two subarrays
        mid = len(array)//2
        left_half = array[:mid]
        right_half = array[mid:]

        # Sort the two halves
        merge_sort(left_half)
        merge_sort(right_half)

        i = j = k = 0

        # Until we reach either end of either left_half or right_half, pick larger among
        # elements left_half and right_half and place them in the correct position at A[p..r]
        while i < len(left_half) and j < len(right_half):
            if left_half[i] < right_half[j]:
                array[k] = left_half[i]
                i += 1
            else:
                array[k] = right_half[j]
                j += 1
            k += 1

        # When we run out of elements in either L or M,
        # pick up the remaining elements and put in A[p..r]
        while i < len(left_half):
            array[k] = left_half[i]
            i += 1
            k += 1

        while j < len(right_half):
            array[k] = right_half[j]
            j += 1
            k += 1

def two_pivot_block_partitioning(array: list[int], start: int, end: int) -> tuple[int, int]:
    """Returns two pivot index that divide the array into three parts using block partitioning method

    Args:
        array (list[int]): Array that will be partitioned
        start (int): The start of the partition
        end (int): The end of the partition

    Returns:
        tuple[int, int]: A tuple of the first and second pivot index
    """
    BLOCK_SIZE = 1024
    if (end <= start):
        return (start, start)
    pivot1 = array[start]
    pivot2 = array[end]
    
    if (pivot1 > pivot2):
        array[start], array[end] = array[end], array[start]
        pivot1, pivot2 = pivot2, pivot1
    
    block = [None] * BLOCK_SIZE
    
    i, j, k = start + 1, start + 1, start + 1
    nump, numq = 0, 0
    
    while k < end:
        t = min(BLOCK_SIZE, end - k)
        for c in range(t):
            block[numq] = c
            numq = numq + (pivot2 >= array[k + c])
        for c in range(numq):
            array[j+c], array[k+block[c]] = array[k+block[c]], array[j+c]
        k += t
        for c in range(numq):
            block[nump] = c
            nump += (pivot1 > array[j + c])
        for c in range(nump):
            array[i], array[j + block[c]] = array[j + block[c]], array[i]
            i += 1
        j += numq
        nump, numq = 0, 0
    
    array[i-1], array[start] = array[start], array[i-1]
    array[j], array[end] = array[end], array[j]
    return (i-1, j)

def two_pivot_block_quicksort(array: list[int], start: int, end: int) -> None:
    if start <= end:
        pivots = two_pivot_block_partitioning(array, start, end)
        two_pivot_block_quicksort(array, start, pivots[0] - 1)
        two_pivot_block_quicksort(array, pivots[0] + 1, pivots[1] - 1)
        two_pivot_block_quicksort(array, pivots[1] + 1, end)

File: SortingAlgo.java
import java.util.Arrays;

public class SortingAlgo {

    /**
     * Sort a given array of int using merge sort algorithm
     * Source: <a href="https://www.programiz.com/dsa/merge-sort">programiz</a>
     * @param array An array of int that want to be sorted
     */
    public static void mergeSort(int[] array) {
        if (array.length > 1) {
            int mid = array.length / 2;
            int[] leftHalf = Arrays.copyOfRange(array, 0, mid);
            int[] rightHalf = Arrays.copyOfRange(array, mid, array.length);

            mergeSort(leftHalf);
            mergeSort(rightHalf);

            int i = 0, j = 0, k = 0;
            while (i < leftHalf.length && j < rightHalf.length) {
                if (leftHalf[i] < rightHalf[j]) {
                    array[k] = leftHalf[i];
                    i++;
                } else {
                    array[k] = rightHalf[j];
                    j++;
                }
                k++;
            }

            while (i < leftHalf.length) {
                array[k] = leftHalf[i];
                i++;
                k++;
            }

            while (j < rightHalf.length) {
                array[k] = rightHalf[j];
                j++;
                k++;
            }
        }
    }

    /**
     * Partition an array from start to end into three part using two pivots.
     * Uses lomuto with block partitioning method.
     * @param array The array that will be partitioned
     * @param start The start index of the subarray
     * @param end The end index of the subarray
     * @param blockSize The size of the block that will be used
     * @return An array with index 0 as the first pivot and index 1 as the
     * second pivot
     */
    public static int[] twoPivotBlockPartitioning(int[] array, int start, int end, int blockSize) {
        if (end <= start) {
            return new int[]{start, start};
        }

        int pivot1 = array[start];
        int pivot2 = array[end];

        if (pivot1 > pivot2) {
            int temp = pivot1;
            pivot1 = pivot2;
            pivot2 = temp;

            temp = array[start];
            array[start] = array[end];
            array[end] = temp;
        }

        int[] block = new int[blockSize];
        int i = start + 1, j = start + 1, k = start + 1;
        int numP = 0, numQ = 0;

        while (k < end) {
            int t = Math.min(blockSize, end - k);

            for (int c = 0; c < t; c++) {
                block[numQ] = c;
                numQ += (pivot2 >= array[k + c]) ? 1 : 0;
            }

            for (int c = 0; c < numQ; c++) {
                int temp = array[j + c];
                array[j + c] = array[k + block[c]];
                array[k + block[c]] = temp;
            }

            k += t;

            for (int c = 0; c < numQ; c++) {
                block[numP] = c;
                numP += (pivot1 > array[j + c]) ? 1 : 0;
            }

            for (int c = 0; c < numP; c++) {
                int temp = array[i];
                array[i] = array[j + block[c]];
                array[j + block[c]] = temp;
                i++;
            }

            j += numQ;
            numP = 0;
            numQ = 0;
        }

        int temp = array[i - 1];
        array[i - 1] = array[start];
        array[start] = temp;

        temp = array[j];
        array[j] = array[end];
        array[end] = temp;

        return new int[]{i - 1, j};
    }

    public static void twoPivotBlockQuickSort(int[] array, int start, int end) {
        if (array.length == 0) {
            return;
        }

        // Berdasarkan paper rujukan, untuk data ukuran 2^26, block size yang
        // optimal adalah 2^10. saya simpulkan bahwa untuk data ukuran 2^n,
        // block size yang optimal adalah 2^(floor(n/3) + 2)
        int blockSize = (int) Math.pow(2, Math.floor(Math.log(array.length) / Math.log(2) / 3 + 2));
        int[][] stack = new int[array.length][2];
        int top = -1;

        // Quicksort dimodifikasi hingga dalam bentuk iteratif (tapi
        // behaviornya tetap sama seperti recursive call) karena untuk data
        // yang terlalu besar, terjadi stack overflow
        stack[++top] = new int[]{start, end};

        while (top >= 0) {
            int[] range = stack[top--];
            start = range[0];
            end = range[1];

            if (start < end) {
                int[] pivots = twoPivotBlockPartitioning(array, start, end, blockSize);
                stack[++top] = new int[]{start, pivots[0] - 1};
                stack[++top] = new int[]{pivots[0] + 1, pivots[1] - 1};
                stack[++top] = new int[]{pivots[1] + 1, end};
            }
        }
    }

    public static void main(String[] args) {
        int[] array = {5, 7, 3, 1, 6, 9};
        twoPivotBlockQuickSort(array, 0, array.length - 1);
        System.out.println(Arrays.toString(array));

        array = new int[]{1, 2, 3, 4, 5};
        mergeSort(array);
        System.out.println(Arrays.toString(array));
    }
}